tvm-tac/src/main/kotlin/org/ton/tac/TacGenerator.kt

```diff
@@ -0,0 +1,237 @@
+package org.ton.tac
+
+import org.ton.bytecode.TvmConstDataInst
+import org.ton.bytecode.TvmContOperandInst
+import org.ton.bytecode.TvmContractCode
+import org.ton.bytecode.TvmDisasmCodeBlock
+import org.ton.bytecode.TvmStackBasicInst
+import org.ton.bytecode.TvmStackComplexInst
+import org.ton.bytecode.extractPrimitiveOperands
+
+data class ContGeneratedInfo(
+    val contInstructions: MutableList<TacInst>,
+    val contArgs: List<TacVar>,
+    val contStackAfter: Stack
+)
+
+object TacVarFactory {
+    private var contVarCounter = 0
+
+    fun nextContVarName(): String {
+        return "var_${contVarCounter++}"
+    }
+}
+
+fun generateTacCodeBlock(
+    codeBlock: TvmDisasmCodeBlock,
+    stack: Stack = Stack(emptyList()),
+    endingFromParentInst: NonStackTacInst? = null,
+    contract: TvmContractCode,
+    calledMethodsSet: MutableSet<MethodId> = mutableSetOf(),
+): ContGeneratedInfo {
+    val tacInstructions = mutableListOf<TacInst>()
+
+    for (inst in codeBlock.instList) {
+        val operands = extractPrimitiveOperands(inst).toMutableMap()
+
+        throwErrorIfBranchesNotTypeVar(inst)
+        val (allBranchesWithSave, endingInst) = checkAllBranchesWithSave(inst.branches, inst.mnemonic)
+
+        if (inst !is TvmStackBasicInst && inst !is TvmStackComplexInst) {
```

Comment:
Also, please keep in mind that we are going to extend this code. Right now, adding instruction with `tuple` or `conditional` stack effect requires a lot of changes in this code. It should be structured better